flink-end-to-end-tests/flink-rest-api-test/src/main/java/org/apache/flink/runtime/rest/tests/RestApiTestSuite.java

```diff
@@ -0,0 +1,363 @@
+import org.apache.flink.runtime.rest.messages.MessageHeaders;
+import org.apache.flink.runtime.rest.messages.job.JobDetailsInfo;
+import org.apache.flink.runtime.rest.messages.taskmanager.TaskManagersInfo;
+import org.apache.flink.runtime.rpc.FatalErrorHandler;
+import org.apache.flink.runtime.webmonitor.retriever.GatewayRetriever;
+import org.apache.flink.runtime.webmonitor.retriever.MetricQueryServiceRetriever;
+import org.apache.flink.runtime.webmonitor.testutils.HttpTestClient;
+import org.apache.flink.util.ConfigurationException;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.flink.shaded.netty4.io.netty.channel.ChannelInboundHandler;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import scala.concurrent.duration.FiniteDuration;
+
+/**
+ * Rest API test suite.
+ */
+public class RestApiTestSuite {
+	private static final FiniteDuration TEST_TIMEOUT = new FiniteDuration(10L, TimeUnit.SECONDS);
+	private static final ObjectMapper MAPPER = new ObjectMapper();
+
+	private static int testSuccessCount = 0;
+	private static int testFailureCount = 0;
+	private static int testSkipCount = 0;
+
+	public static void main(String[] args) throws Exception {
+
+		ParameterTool params = ParameterTool.fromArgs(args);
+		final String host = params.get("host", "localhost");
+		final int port = params.getInt("port", 8081);
+		final HttpTestClient httpClient = new HttpTestClient(host, port);
+
+		// Validate Flink cluster is running
+		JobIdsWithStatusOverview jobOverview = getJobOverview(httpClient);
+
+		// Get necessary parameters for testing GET endpoints
+		Map<String, String> parameterMap = getParameterMaps(httpClient, jobOverview);
+
+		// Get list of endpoints
+		List<MessageHeaders> specs = new E2ETestDispatcherRestEndpoint().getSpecs();
+		specs.forEach(spec -> testMonitoringEndpointSpecs(httpClient, spec, parameterMap));
+
+		if (testFailureCount != 0) {
+			throw new RuntimeException("There are test failures. Success: " + testSuccessCount +
+				" Failures: " + testFailureCount + " Skipped: " + testSkipCount);
+		}
+	}
+
+	@SuppressWarnings("ConstantConditions")
+	private static Map<String, String> getParameterMaps(HttpTestClient httpClient,
+		JobIdsWithStatusOverview jobOverview) throws InterruptedException, IOException, TimeoutException {
+		// Get necessary parameters used for all REST API testings.
+		final Map<String, String> parameterMap = new HashMap<>();
+		Preconditions.checkState(jobOverview.getJobsWithStatus().stream()
```

Comment:
May be better to make this an assertion, effectively we are relying on the REST API functionality here already.